This tutorial covers some of the basic functionality of the sigpyproc package. For a guide on how to extend the
package, see the

## Getting started

For test puproses, few small size filterbank files are included in the `/tests/data/` directory of the `sigpyproc` package.

### Loading data into sigpyproc

Lets start by loading our filterbank file into sigpyproc. To do this, we require the :class:`~sigpyproc.readers.FilReader` class from the :automodule:`sigpyproc.readers` module.

In [1]:
import numpy as np
from rich.pretty import Pretty
from sigpyproc.readers import FilReader

In [2]:
myFil = FilReader("../../tests/data/tutorial.fil")

In [3]:
myFil

`myFil` now contains an instance of the :class:`sigpyproc.readers.FilReader` class. We can access obervational
metadata through the `myFil.header` attribute:

In [4]:
Pretty(myFil.header)

Header(
    filename='../../tests/data/tutorial.fil',
    data_type='filterbank',
    nchans=64,
    foff=-1.09,
    fch1=1510.0,
    nbits=2,
    tsamp=0.00032,
    tstart=50000.0,
    nsamples=187520,
    nifs=1,
    coord=<SkyCoord (ICRS): (ra, dec) in deg
    (0., 0.)>,
    azimuth=<Angle 0. deg>,
    zenith=<Angle 0. deg>,
    telescope='Fake',
    backend='FAKE',
    source='P: 250.000000000000 ms, DM: 30.000',
    frame='topocentric',
    ibeam=0,
    nbeams=0,
    dm=0,
    period=0,
    accel=0,
    signed=False,
    rawdatafile=None,
    hdrlens=[244],
    datalens=[3000320],
    filenames=['../../tests/data/tutorial.fil'],
    nsamples_files=[187520],
    tstart_files=[50000.0]
)

where

In [5]:
type(myFil.header)

sigpyproc.header.Header

All values stored in the `myFil.header` attribute may be accessed as attributes, i.e.:

In [6]:
myFil.header.nchans

64

Now that we know how to load a file into `sigpyproc`, let’s look at at doing something with the loaded data.

### Dedispersing the data

One of the most used techniques in pulsar processing is dedispersion, wherein we add or remove frequency dependent
time delays to the data.

To dedisperse our `myFil` instance, we simply call the dedisperse method:

In [7]:
myTim = myFil.dedisperse(30)

Output()

In [8]:
myTim

TimeSeries([108., 100., 102., ..., 105., 111., 107.], dtype=float32)

In [9]:
type(myTim)

sigpyproc.timeseries.TimeSeries

Here we have dedispersed to a DM of 30 pc cm$^{-3}$ with the result being an instance of the
:autoclass:`sigpyproc.timeseries.TimeSeries` class, which we have called `myTim`.

The :autoclass:`sigpyproc.timeseries.TimeSeries` class in a subclass of `numpy.ndarray`, and is capable of using
all standard numpy functions. For example:

In [10]:
myTim.sum()

TimeSeries(19636992., dtype=float32)

In [11]:
myTim.max()

TimeSeries(121., dtype=float32)

In [12]:
myTim.min()

TimeSeries(88., dtype=float32)

In [13]:
np.median(myTim)

TimeSeries(105., dtype=float32)

The use of `numpy.ndarray` subclasses is important in allowing sigpyproc to easily interface with many 3rd party
python libraries.

### Performing a Fourier transform

To perform a discrete fourier transform of the data contained in the `myTim` instance we may invoke the `myTim.rFFT`
method.

In [14]:
myFS = myTim.rfft()

In [15]:
type(myFS)

sigpyproc.fourierseries.FourierSeries

In [16]:
myFS

FourierSeries([ 1.9636884e+07   +0.j     , -2.9424850e+02 +429.87863j,
                5.4652838e+02 -577.44696j, ...,
               -1.3942198e+03+1670.1677j , -1.2117781e+03-1779.332j  ,
               -2.7670000e+03   +0.j     ], dtype=complex64)

The :autoclass:`sigpyproc.fourierseries.FourierSeries` is also a subclass of `numpy.ndarray`, where array elements are `numpy.complex64`.

Using the `remove_rednoise` method of `myFS`, we can de-redden the Fourier series:

In [17]:
myFS_red = myFS.remove_rednoise()

In [18]:
myFS_red

FourierSeries([ 1.       +0.j      ,        nan     +nanj,
                      nan     +nanj, ..., -0.8724377+1.045113j,
               -0.7582742-1.113423j, -1.7314595+0.j      ],
              dtype=complex64)

with the dereddened fourier series, we can now form the power spectrum of the observation:

In [19]:
mySpec = myFS_red.form_spec(interpolated=True)

In [20]:
mySpec

PowerSpectrum([1.       ,       nan,       nan, ..., 1.3613995, 1.5284487,
               1.7314595], dtype=float32)

Here we have set the `interpolated` flag to True, causing the `formSpec` function to perform nearest bin interpolation.

`mySpec` contains several convenience methods to help with navigating the power spectrum. For instance:

In [21]:
mySpec.period2bin(0.25)

240

In [22]:
mySpec.freq2bin(5.0)

300

We can also perofrm Lyne-Ashworth harmonic folding to an arbitrary number of harmonics:

In [23]:
folds = mySpec.harmonic_fold(5)

In [24]:
folds

[PowerSpectrum([1.       ,       nan,       nan, ..., 2.7979643, 2.9650135,
                1.7314595], dtype=float32),
 PowerSpectrum([1.       ,       nan,       nan, ..., 2.7979643, 2.9650135,
                1.7314595], dtype=float32),
 PowerSpectrum([1.       ,       nan,       nan, ..., 2.7979643, 2.9650135,
                1.7314595], dtype=float32),
 PowerSpectrum([1.       ,       nan,       nan, ..., 2.7979643, 2.9650135,
                1.7314595], dtype=float32),
 PowerSpectrum([1.       ,       nan,       nan, ..., 2.7979643, 2.9650135,
                1.7314595], dtype=float32)]

Where the variable `folds` is a python list containing each of the requested harmonic folds.

### Folding data

Both the :autoclass:`sigpyproc.timeseries.TimeSeries` and the :autoclass:`sigpyproc.fourierseries.FourierSeries`
have methods to phase fold their data. Using our earlier myFil instance, we will fold our filterbank file with a period
of 250 ms and a DM of pc cm$^{-3}$ and acceleration of 0 ms$^{-2}$.

In [25]:
myFold = myFil.fold(0.25,30.,accel=0,nbins=128,nints=32,nbands=16)

Output()

In [26]:
type(myFold)

sigpyproc.foldedcube.FoldedData

In [27]:
myFold.shape

(32, 16, 128)

The the :autoclass:`sigpyproc.foldedcube.FoldedData` has several functions to enable simple slicing and summing of
the folded data cube. These include:

* `getSubband`: select all data in a single frequency band
* `getSubint`: select all data in a single subintegration
* `getFreqPhase`: sum the data in the time axis
* `getTimePhase`: sum the data in the frequency axis
*`getProfile`: get the pulse profile of the fold

We can also tweak the DM and period of the fold using the `updateParams` method:

In [28]:
myFold.update_dm(dm=100)
myFold.update_period(period=0.2502)

### Tips and tricks

There are several tips and tricks to help speed up `sigpyproc` and also make it more user friendly. For people who are
familiar with Python and IPython these will be old news, but for newbies these may be of use.

**Tab completion**: One of the many nice things about IPython is that it allows for tab completion:

In [30]:
myFil   # then press tab

**Docstrings**: by using question marks or double question marks we can access both information about a function and
its raw source:

In [31]:
myFil.downsample?

Note that all docstrings are written in `numpydoc`. This is to facilitate automatic documentation creation with the
Sphinx package.

**Chaining**: The ability to chain together methods, combined with history recall in IPython means that it is simple to
condense a sigpyproc request into a single line:

In [33]:
spectrum = FilReader("../../tests/data/tutorial.fil").collapse().rfft().remove_rednoise().form_spec(True)

Output()

Here we create a `FilReader` instance, which is then collapsed in frequency, FFT’d, cleaned of rednoise and interpolated to form a power spectrum. In the intrests of readability, this is not always a good idea, however for testing code
quickly, it is invaluable.